mira/metamodel/templates.py

```diff
@@ -134,6 +284,177 @@ def get_json_schema():
+        if key == "type":
+            continue
+
+        if isinstance(value, Concept):
+            other_concept: Concept = other_dict[key]
+            if not value.is_equal_to(other_concept, with_context=with_context):
+                return False
+
+        elif isinstance(value, list):
+            if not all(i1 == i2 for i1, i2 in zip(value, other_dict[key])):
+                return False
+        else:
+            raise NotImplementedError(
+                f"No comparison implemented for type {type(value)} for Template"
+            )
+    return True
+
+
+def assert_concept_context_refinement(refined_concept: Concept, other_concept: Concept) -> bool:
+    """Check if one Concept's context is a refinement of another Concept's
+
+    Special case:
+    - Both contexts are empty => special case of equal context => False
+
+    Parameters
+    ----------
+    refined_concept :
+        The assumed *more* detailed Concept
+    other_concept :
+        The assumed *less* detailed Concept
+
+    Returns
+    -------
+    :
+        True if the Concept `refined_concept` truly is strictly more detailed
+        than `other_concept`
+    """
+    refined_context = refined_concept.context
+    other_context = other_concept.context
+    # 1. False if no context for both
+    if len(refined_context) == 0 and len(other_context) == 0:
+        return False
+    # 2. True if refined concept has context and the other one not
+    elif len(refined_context) > 0 and len(other_context) == 0:
+        return True
+    # 3. False if refined concept does not have context and the other does
+    elif len(refined_context) == 0 and len(other_context) > 0:
+        return False
+    # 4. Both have context
+    else:
+        # 1. Exactly equal context keys -> False
+        # 2. False if refined Concept context is a subset of other context
+        #
+        # NOTE: issubset is not strict, i.e. is True for equal sets, therefore
+        # we need to check for refined.issubset(other) first to be sure that
+        # cases 1. and 2. are ruled out when 3. is evaluated
+        if set(refined_context.keys()).issubset(other_context.keys()):
+            return False
+
+        # 3. Other Concept context is a subset; check equality for the matches
+        elif set(other_context.keys()).issubset(refined_context):
+            for other_context_key, other_context_value in other_context.items():
+                if refined_context[other_context_key] != other_context_value:
+                    return False
+
+        # 4. Both Concepts have context, but they are different -> cannot be a
+        #    refinement -> False
+        elif set(other_context.keys()).symmetric_difference(set(refined_context.keys())):
+            return False
+
+        # 5. All cases should be covered, but in case something is missing
+        else:
+            raise ValueError("Unhandled logic, missing at least one logical option")
+
+    return True
+
+
+def get_relations_web(
```

Comment:
I think we should move this to the `dkg` module, maybe in a file called `web_client.py`. The template classes shouldn't explicitly use / depend on this function. Rather, they should take as input a function with a simpler interface, taking a source and a target as input and returning True/False as output if a refinement relationship exists. When calling refinement over some templates, this function can then be supplied by outside code.